### PR TITLE
[SPARK-17914][SQL] Fix parsing of timestamp strings with nanoseconds

### PR DESCRIPTION
**Pull request by aokolnychyi**
The PR contains a tiny change to fix the way Spark parses string literals into timestamps. Currently, some timestamps that contain nanoseconds are corrupted during the conversion from internal UTF8Strings into the internal representation of timestamps.

Consider the following example:
```
spark.sql("SELECT cast('2015-01-02 00:00:00.000000001' as TIMESTAMP)").show(false)
+------------------------------------------------+
|CAST(2015-01-02 00:00:00.000000001 AS TIMESTAMP)|
+------------------------------------------------+
|2015-01-02 00:00:00.000001                      |
+------------------------------------------------+
```

The fix was tested with existing tests. Also, there is a new test to cover cases that did not work previously.

### CODE REVIEW

**Comment by ueshin** [comment]
Looks like [DateTimeUtils.scala#L411-L414](https://github.com/aokolnychyi/spark/blob/2f232a7bda28fb42759ee35923044f886a1ff19e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala#L411-L414) is for the same purpose but not enough.
I guess we can remove these lines now.

**Comment by aokolnychyi** [comment]
@ueshin good point, thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -399,13 +399,13 @@ object DateTimeUtils {
       digitsMilli += 1
     }
 
-    if (!justTime && isInvalidDate(segments(0), segments(1), segments(2))) {
-      return None
+    while (digitsMilli > 6) {
```
Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: add a comment indicating we are truncating the microsecond part and its lossy?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 2]
```diff
@@ -32,7 +32,7 @@ import org.apache.spark.unsafe.types.UTF8String
  * Helper functions for converting between internal and external date and time representations.
  * Dates are exposed externally as java.sql.Date and are represented internally as the number of
  * dates since the Unix epoch (1970-01-01). Timestamps are exposed externally as java.sql.Timestamp
- * and are stored internally as longs, which are capable of storing timestamps with 100 nanosecond
+ * and are stored internally as longs, which are capable of storing timestamps with microsecond
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: 100 ns is different from micro, isn't it?

Comment (aokolnychyi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Sure, but the previous comment, which was introduced in [this](https://github.com/apache/spark/commit/6b7f2ceafdcbb014791909747c2210b527305df9) commit, was no longer correct. The logic was changed in [this](https://github.com/apache/spark/commit/a290814877308c6fa9b0f78b1a81145db7651ca4) commit and now it is up to microseconds.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 1 above)
Comment (aokolnychyi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: @wzhfy done

**Comment by wzhfy** [comment]
LGTM

**Comment by aokolnychyi** [comment]
@wzhfy @rxin @ueshin can someone, please, merge this?

**Comment by ueshin** [comment]
Thanks! merging to master/2.2.

**Pull request closed without merging**